<a href="https://colab.research.google.com/github/MochamadPhilliaWibowo/Machine-Learning-Class/blob/main/Tugas%20Week%2013/BAB%202/Behind_the_pipeline_(PyTorch).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Mochamad Phillia Wibowo

NIM : 1103204191

Kelas : Machine Learning (TK-44-G04)

Tugas 13 NLP HUGGING FACE (BAB 2)

# Behind the pipeline (PyTorch)

Install the Transformers, Datasets, and Evaluate libraries to run this notebook.

In [1]:
!pip install datasets evaluate transformers[sentencepiece]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 542.1/542.1 kB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 2.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 2.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 2.6 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.31.0
    Uninstalling requests-2.31.0:
      Successfully uninstalled requests-2.31.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.31.0, but you have requests 2.32.3 which is incompatible.


Code di atas adalah perintah untuk menginstal beberapa pustaka Python menggunakan pip, manajer paket standar untuk Python. Perintah ini melakukan instalasi tiga pustaka:

datasets: Ini adalah pustaka yang menyediakan akses mudah ke berbagai kumpulan data untuk tugas-tugas pembelajaran mesin dan pemrosesan bahasa alami.

evaluate: Ini mungkin adalah pustaka khusus yang digunakan untuk mengevaluasi model-model pembelajaran mesin atau pemrosesan bahasa alami, tetapi tanpa informasi lebih lanjut, sulit untuk mengkonfirmasi.

transformers[sentencepiece]: Ini adalah pustaka yang populer digunakan untuk pemrosesan bahasa alami, terutama dalam bidang pemodelan bahasa dan transfer pembelajaran. Opsi [sentencepiece] menunjukkan bahwa instalasi juga akan mencakup dukungan untuk SentencePiece, sebuah pustaka tokenisasi teks multibahasa.

Dengan menjalankan perintah ini, pustaka-pustaka ini akan diinstal dan siap digunakan dalam pengembangan aplikasi Python Anda.

In [2]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")
classifier(
    [
        "I've been waiting for a HuggingFace course my whole life.",
        "I hate this so much!",
    ]
)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that aut

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

[{'label': 'POSITIVE', 'score': 0.9598048329353333},
 {'label': 'NEGATIVE', 'score': 0.9994558691978455}]

1. from transformers import pipeline: Baris ini mengimpor modul pipeline dari pustaka Transformers. Transformers adalah pustaka yang menyediakan berbagai alat untuk pemrosesan bahasa alami (Natural Language Processing/NLP), termasuk model untuk analisis sentimen.

2. classifier = pipeline("sentiment-analysis"): Baris ini membuat objek classifier yang merupakan sebuah pipeline untuk analisis sentimen. Kita menggunakan metode pipeline() dan menentukan "sentiment-analysis" sebagai argumennya, yang mengarahkan pipeline untuk menggunakan model yang telah dilatih untuk analisis sentimen.

3. classifier([...]): Baris ini memanggil metode classifier yang telah dibuat sebelumnya untuk melakukan analisis sentimen pada dua kalimat yang diberikan. Dua kalimat yang diberikan dimasukkan dalam sebuah daftar (list).

a. "I've been waiting for a HuggingFace course my whole life.": Kalimat pertama adalah kalimat positif, yang mengungkapkan antusiasme terhadap kursus HuggingFace.

b. "I hate this so much!": Kalimat kedua adalah kalimat negatif, yang menyatakan ketidakpuasan.

Setelah kode dijalankan, outputnya akan berupa hasil analisis sentimen untuk masing-masing kalimat, yang mungkin berupa label sentimen (misalnya: "positive", "negative", "neutral") beserta probabilitasnya.

In [3]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Baris pertama mengimpor AutoTokenizer dari pustaka Transformers. AutoTokenizer adalah kelas yang digunakan untuk menginisialisasi dan memuat tokenizer yang sesuai dengan model yang akan digunakan.

Baris kedua mendefinisikan variabel checkpoint yang berisi string yang menunjukkan model yang akan digunakan. Dalam hal ini, model yang dipilih adalah "distilbert-base-uncased-finetuned-sst-2-english". Model ini adalah model DistilBERT yang telah di-fineturn pada tugas analisis sentimen (SST-2) dalam bahasa Inggris.

Baris ketiga menginisialisasi tokenizer dengan menggunakan metode from_pretrained() dari kelas AutoTokenizer. Metode ini memuat tokenizer yang sesuai dengan model yang ditentukan dalam variabel checkpoint. Sebagai hasilnya, variabel tokenizer akan berisi tokenizer yang dapat digunakan untuk memproses teks dalam model "distilbert-base-uncased-finetuned-sst-2-english".

In [4]:
raw_inputs = [
    "I've been waiting for a HuggingFace course my whole life.",
    "I hate this so much!",
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}


Baris pertama mendefinisikan sebuah daftar yang disebut raw_inputs. Daftar ini berisi dua string, masing-masing adalah kalimat yang akan diproses menggunakan tokenizer yang telah diinisialisasi sebelumnya.

Baris kedua menggunakan tokenizer untuk memproses raw_inputs. Tokenizer menggunakan metode __call__() untuk melakukan tokenisasi terhadap list teks yang diberikan. Argumen-argumen yang diberikan kepada tokenizer adalah sebagai berikut:

- raw_inputs: Daftar teks yang akan diproses.

- padding=True: Menentukan apakah token harus dipad agar memiliki panjang yang seragam. Dalam kasus ini, karena padding=True, tokenizer akan menambahkan padding ke teks agar semua token memiliki panjang yang sama.

- truncation=True: Menentukan apakah teks harus dipotong agar memiliki panjang maksimum yang ditentukan oleh tokenizer. Dalam kasus ini, karena truncation=True, tokenizer akan memotong teks yang terlalu panjang agar sesuai dengan panjang maksimum yang diperbolehkan oleh model.

- return_tensors="pt": Menentukan format tensor yang akan dikembalikan. Dalam hal ini, "pt" berarti tokenizer akan mengembalikan tensor PyTorch.
Hasil dari pemrosesan tersebut disimpan dalam variabel inputs.

Baris ketiga mencetak isi dari variabel inputs, yang berisi tensor yang telah diproses oleh tokenizer. Outputnya adalah tensor PyTorch yang berisi token-token dari kedua kalimat, termasuk token khusus seperti token untuk padding dan token khusus untuk awalan dan akhir teks.

1. 'input_ids': Ini adalah token-token yang dihasilkan dari tokenisasi teks input. Dua kalimat telah diubah menjadi urutan token-token. Setiap token mewakili kata atau bagian dari kata dalam teks, dan angka-angka yang ditunjukkan adalah ID dari masing-masing token dalam kamus token yang digunakan oleh model. Angka 101 adalah ID untuk token [CLS], yang menandakan awal dari teks, dan angka 102 adalah ID untuk token [SEP], yang menandakan akhir dari teks. Angka 0 menandakan token yang ditambahkan sebagai padding.

2. 'attention_mask': Ini adalah tensor yang menunjukkan bagian mana dari input yang sebenarnya berisi informasi yang relevan, dan bagian mana yang hanya berisi padding. Nilai 1 menunjukkan token-token yang sebenarnya, sedangkan nilai 0 menunjukkan token-token yang hanya padding dan seharusnya diabaikan oleh model.

In [6]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModel.from_pretrained(checkpoint)

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Baris pertama mengimpor kelas AutoModel dari pustaka Transformers. AutoModel adalah kelas yang digunakan untuk menginisialisasi dan memuat model yang sesuai dengan tugas yang akan dilakukan. Ini memungkinkan kita untuk menggunakan berbagai model dengan cara yang seragam.

Baris kedua mendefinisikan variabel checkpoint yang berisi string yang menunjukkan model yang akan digunakan. Dalam kasus ini, model yang dipilih adalah "distilbert-base-uncased-finetuned-sst-2-english". Ini adalah model DistilBERT yang telah di-fineturn pada tugas analisis sentimen (SST-2) dalam bahasa Inggris.

Baris ketiga menginisialisasi model dengan menggunakan metode from_pretrained() dari kelas AutoModel. Metode ini memuat model yang sesuai dengan model yang ditentukan dalam variabel checkpoint. Sebagai hasilnya, variabel model akan berisi model yang siap digunakan untuk analisis sentimen atau tugas lain yang sesuai dengan model tersebut.

In [7]:
outputs = model(**inputs)
print(outputs.last_hidden_state.shape)

torch.Size([2, 16, 768])


1. outputs = model(**inputs): Baris ini melakukan inferensi menggunakan model yang telah diinisialisasi sebelumnya. Model menerima input dalam bentuk inputs, yang merupakan keluaran dari proses tokenisasi sebelumnya. **inputs mengembalikan argumen-argumen yang dibuat dalam format yang dapat diterima oleh model, yang pada dasarnya berupa tensor PyTorch.

2. print(outputs.last_hidden_state.shape): Baris ini mencetak bentuk dari keluaran terakhir (last_hidden_state) dari model. Keluaran ini biasanya merupakan representasi dari teks input di dalam ruang laten model. Bentuknya adalah tuple yang menunjukkan dimensi dari tensor, dalam hal ini, (2, 16, 768). Ini menunjukkan bahwa ada dua teks input dalam batch (2), masing-masing dengan 16 token (termasuk token-padding), dan masing-masing token direpresentasikan dalam ruang fitur 768 dimensi.

In [8]:
from transformers import AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
outputs = model(**inputs)

1. from transformers import AutoModelForSequenceClassification: Baris ini mengimpor kelas AutoModelForSequenceClassification dari pustaka Transformers. AutoModelForSequenceClassification adalah kelas yang digunakan untuk menginisialisasi dan memuat model yang sudah di-pretrained untuk tugas klasifikasi urutan teks.

2. checkpoint = "distilbert-base-uncased-finetuned-sst-2-english": Baris ini mendefinisikan variabel checkpoint yang berisi string yang menunjukkan model yang akan digunakan. Dalam hal ini, model yang dipilih adalah "distilbert-base-uncased-finetuned-sst-2-english". Ini adalah model DistilBERT yang telah di-fineturn pada tugas analisis sentimen (SST-2) dalam bahasa Inggris.

3. model = AutoModelForSequenceClassification.from_pretrained(checkpoint): Baris ini memuat model dari checkpoint yang telah ditentukan sebelumnya. Metode from_pretrained() digunakan untuk memuat model yang sesuai dengan model yang ditentukan dalam variabel checkpoint. Sebagai hasilnya, variabel model akan berisi model yang siap digunakan untuk melakukan klasifikasi urutan teks.

4. outputs = model(**inputs): Baris ini menggunakan model yang telah diinisialisasi untuk melakukan inferensi pada data input yang telah diproses sebelumnya. Input yang diberikan adalah inputs, yang merupakan keluaran dari proses tokenisasi sebelumnya. **inputs mengembalikan argumen-argumen dalam format yang dapat diterima oleh model, yang pada dasarnya berupa tensor PyTorch. Hasil inferensi disimpan dalam variabel outputs.

In [9]:
print(outputs.logits.shape)

torch.Size([2, 2])


Baris ini mencetak bentuk dari keluaran logits dari model. Logits adalah nilai-nilai mentah sebelum melewati fungsi aktivasi yang menentukan kelas-kelas yang diprediksi oleh model. Mari kita bahas:

1. outputs.logits: Ini adalah atribut dari objek outputs yang berisi nilai-nilai logits dari model. Dalam konteks klasifikasi urutan teks, ini adalah representasi dari prediksi kelas yang dibuat oleh model untuk setiap sampel input.

2. .shape: Ini adalah metode yang digunakan untuk mendapatkan bentuk dari tensor. Metode ini mengembalikan tuple yang menunjukkan dimensi dari tensor tersebut.

In [10]:
print(outputs.logits)

tensor([[-1.5607,  1.6123],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>)


Cetakan ini akan menampilkan nilai-nilai logits yang dihasilkan oleh model untuk setiap sampel input. Logits ini menunjukkan nilai mentah sebelum melewati fungsi aktivasi yang akan menentukan kelas prediksi. Ini bergantung pada jumlah kelas yang ditentukan oleh model dan tugas klasifikasi yang dilakukan.

Keluaran ini adalah tensor PyTorch yang berisi nilai-nilai logits yang dihasilkan oleh model untuk dua sampel input. Dalam hal ini, ada dua sampel input, dan untuk setiap sampel, model menghasilkan dua nilai logits. Setiap nilai dalam tensor mewakili nilai mentah atau skor untuk setiap kelas dalam tugas klasifikasi. Misalnya,

untuk baris pertama dari tensor:

1. Nilai pertama adalah -1.5607, yang menunjukkan skor untuk kelas pertama.

2. Nilai kedua adalah 1.6123, yang menunjukkan skor untuk kelas kedua.

Untuk baris kedua dari tensor:

1. Nilai pertama adalah 4.1692, yang menunjukkan skor untuk kelas pertama.

2. Nilai kedua adalah -3.3464, yang menunjukkan skor untuk kelas kedua.
Dalam tugas klasifikasi, nilai-nilai ini akan diubah menjadi probabilitas kelas menggunakan fungsi softmax atau fungsi aktivasi yang sesuai, yang dapat memberikan prediksi kelas yang akhir.

In [11]:
import torch

predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
print(predictions)

tensor([[4.0195e-02, 9.5980e-01],
        [9.9946e-01, 5.4418e-04]], grad_fn=<SoftmaxBackward0>)


1. import torch: Baris ini mengimpor modul PyTorch, yang akan digunakan untuk melakukan operasi matematika pada tensor.

2. predictions = torch.nn.functional.softmax(outputs.logits, dim=-1): Baris ini menggunakan fungsi softmax dari modul PyTorch untuk mengonversi logits yang dihasilkan oleh model menjadi probabilitas untuk setiap kelas. Argumen outputs.logits adalah tensor yang berisi logits yang dihasilkan oleh model. Argumen kedua, dim=-1, menunjukkan dimensi mana yang akan dijadikan acuan untuk menghitung softmax. Di sini, -1 mengacu pada dimensi terakhir dari tensor, yang akan dijadikan acuan untuk menghitung softmax untuk setiap sampel input secara terpisah.

3. print(predictions): Baris ini mencetak hasil dari operasi softmax, yang merupakan probabilitas untuk setiap kelas untuk setiap sampel input. Probabilitas ini merupakan hasil prediksi model setelah melewati fungsi softmax. Probabilitas tersebut akan berada dalam rentang 0 hingga 1, dan jumlah probabilitas untuk setiap sampel akan sama dengan 1.0.

In [12]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}

model.config.id2label adalah atribut dari objek model yang menyediakan pemetaan dari ID kelas ke label kelas. Dalam konteks klasifikasi, ini berguna untuk mengetahui kelas yang sesuai dengan setiap nilai dalam tensor logits atau probabilitas yang dihasilkan oleh model.

Ini berarti kelas dengan ID 0 diinterpretasikan sebagai kelas "negative" dan kelas dengan ID 1 diinterpretasikan sebagai kelas "positive".

Dengan demikian, ketika kita memiliki hasil prediksi dalam bentuk tensor logits atau probabilitas, kita dapat menggunakan model.config.id2label untuk mendapatkan label yang sesuai untuk setiap nilai dalam tensor tersebut. Ini membantu dalam menafsirkan hasil prediksi model.